Consumer Scheduling

kylehuntsman opened the pull request:
### Overview

Added scheduling logic for the consumer to consume messages during a configurable time window. Also added the moment library.

kylehuntsman committed:
Updated sqs-consumer.js

Added scheduling logic for the consumer to consume messages during a configurable time window. Added the moment library.

kylehuntsman committed:
Updated sqs-consumer.js

Changed scheduling consuming logic to use duration instead of end time. Added unit tests for the scheduled consuming logic.

kylehuntsman committed:
Updated sqs-consumer.js

Renamed scheduling function

sukrit007 commented on lib/sqs/sqs-base.js:
```diff
@@ -24,6 +24,14 @@ const DEFAULT_BASE_AWS_EXT_CONFIG = {
+        scheduler: {
+          scheduled: false,
+          start: '00:00:00',
+          duration: 2 * 60 * 60, // 2hrs
```
Suggestion:  Use human readable duration

kylehuntsman replied:
Added [`human-interval`](https://www.npmjs.com/package/human-interval) package from npm.

sukrit007 replied:
LGTM!

sukrit007 commented on lib/sqs/sqs-base.js:
```diff
@@ -24,6 +24,14 @@ const DEFAULT_BASE_AWS_EXT_CONFIG = {
+          scheduled: false,
+          start: '00:00:00',
+          duration: 2 * 60 * 60, // 2hrs
+          maxWaitSeconds: 6 * 60 * 60 // 6hrs
```
Should this be maxVisibilityTimeout ?  I did not see reference for maxWaitSeconds in the code

kylehuntsman replied:
Yes it should

sukrit007 replied:
LGTM!

sukrit007 commented on test/unit/sqs/sqs-consumer-spec.js:
```diff
@@ -244,10 +248,156 @@ describe('SqsConsumer', () => {
+          .promise()
+          .then(data => consumer._scheduledConsuming(data))
+          .then(data => {
+            data.should.deep.equal({});
```
Suggestion: Would be great if we can also assert that message visibility is set to to required duration

sukrit007 replied:
LGTM!

sukrit007 reviewed (commented):
@FunnyGopher I like your approach. Just added minor comments.  Let me know if you need clarifications with any of those

kylehuntsman committed:
Updated sqs-consumer.js

Can now return scheduler details

kylehuntsman committed:
Updated sqs-sconsumer.js and sqs-base.js

Added human readbale duration values in the config for the consumer scheduler. Added the human-interval package. Added tests for visibility timeout value.

kylehuntsman committed:
Updated package.json

Added human-interval package

kylehuntsman committed:
Updated sqs-consumer.js and utils.js

Added function for getting visibility timeout in the consumer. Removed visibility timeout function from utils.

sukrit007 commented on test/unit/sqs/sqs-consumer-spec.js:
```diff
@@ -244,10 +248,234 @@ describe('SqsConsumer', () => {
+
+      it('resolves the original data', () => {
+        sqs.receiveMessage.returns({ promise: () => Promise.resolve(messages)});
+        return sqs.receiveMessage()
```
Not sure I follow this line.  Line 290 mocks the receiveMessage method and the test is calling mocked method receiveMessage.   
- Is there any reason we need to call receiveMessage when we are just testing _scheduledConsuming ?

Shouldn't the test be only calling consumer._scheduledConsuming(messages) ?

kylehuntsman replied:
Totally agree. All the _scheduledConsuming tests actually did this. I'll fix them.

sukrit007 replied:
LGTM!

kylehuntsman committed:
Updated sqs-consumer-spec.js unit tests

Implmented fixes for comment https://github.com/meltmedia/aws-sdk-js-ext/pull/2#discussion_r80928474
